# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/windows/CompositorWidgetParent.h
# Commit: 71e63781b38c
# Full Hash: 71e63781b38cbde024a9057553cae96c98e5ecb2
# Author: Chris Martin <cmartin@mozilla.com>
# Date: 2020-02-13 18:21:33
# Regressor Bug: 1604412
# File Overlap Count: 2
# Description:
#   Bug 1604412 - Enable remote backbuffer GDI compositing r=jmathies,jld
#   
#   This change adds new "remote backbuffer" logic when compositing without
#   HW acceleration on Windows (IE compositing through Cairo using the Win32
#   GDI)
# ==============================================================================

diff -r f2088b5ae50d -r 71e63781b38c widget/windows/CompositorWidgetParent.h
--- a/widget/windows/CompositorWidgetParent.h	Thu Feb 13 07:13:50 2020 +0200
+++ b/widget/windows/CompositorWidgetParent.h	Thu Feb 13 03:59:13 2020 +0000
@@ -13,6 +13,10 @@
 namespace mozilla {
 namespace widget {
 
+namespace remote_backbuffer {
+class Client;
+}
+
 class CompositorWidgetParent final : public PCompositorWidgetParent,
                                      public WinCompositorWidget {
  public:
@@ -20,6 +24,8 @@
                                   const layers::CompositorOptions& aOptions);
   ~CompositorWidgetParent() override;
 
+  bool Initialize(const RemoteBackbufferHandles& aRemoteHandles);
+
   bool PreRender(WidgetRenderingContext*) override;
   void PostRender(WidgetRenderingContext*) override;
   already_AddRefed<gfx::DrawTarget> StartRemoteDrawing() override;
@@ -35,6 +41,8 @@
 
   bool HasGlass() const override;
 
+  mozilla::ipc::IPCResult RecvInitialize(
+      const RemoteBackbufferHandles& aRemoteHandles) override;
   mozilla::ipc::IPCResult RecvEnterPresentLock() override;
   mozilla::ipc::IPCResult RecvLeavePresentLock() override;
   mozilla::ipc::IPCResult RecvUpdateTransparency(
@@ -52,16 +60,6 @@
   void SetRootLayerTreeID(const layers::LayersId& aRootLayerTreeId) override;
 
  private:
-  bool RedrawTransparentWindow();
-
-  // Ensure that a transparent surface exists, then return it.
-  RefPtr<gfxASurface> EnsureTransparentSurface();
-
-  HDC GetWindowSurface();
-  void FreeWindowSurface(HDC dc);
-
-  void CreateTransparentSurface(const gfx::IntSize& aSize);
-
   RefPtr<VsyncObserver> mVsyncObserver;
   Maybe<layers::LayersId> mRootLayerTreeID;
 
@@ -70,17 +68,13 @@
   gfx::CriticalSection mPresentLock;
 
   // Transparency handling.
-  mozilla::Mutex mTransparentSurfaceLock;
   mozilla::Atomic<nsTransparencyMode, MemoryOrdering::Relaxed>
       mTransparencyMode;
-  RefPtr<gfxASurface> mTransparentSurface;
-  HDC mMemoryDC;
-  HDC mCompositeDC;
 
   // Locked back buffer of BasicCompositor
   uint8_t* mLockedBackBufferData;
 
-  bool mNotDeferEndRemoteDrawing;
+  std::unique_ptr<remote_backbuffer::Client> mRemoteBackbufferClient;
 };
 
 }  // namespace widget